### PR TITLE
change `UseFragmentResult` to union type

### PR DESCRIPTION
**Pull request by phryneas**
This should be merged after #10764 and is the second half of suggestions from #10760 and would address #10650.

It also adds a test to make sure the types actually reflect the real behaviour.

### Checklist:

- [x] If this PR contains changes to the library itself (not necessary for e.g. docs updates), please include a changeset (see [CONTRIBUTING.md](https://github.com/apollographql/apollo-client/blob/main/CONTRIBUTING.md#changesets))
- [ ] If this PR is a new feature, please reference an issue where a consensus about the design was reached (not necessary for small changes)
- [x] Make sure all of the significant new logic is covered by tests

### CODE REVIEW

File: src/react/hooks/useFragment.ts [hunk 1]
```diff
@@ -115,7 +121,7 @@ function diffToResult<TData>(
   diff: Cache.DiffResult<TData>,
 ): UseFragmentResult<TData> {
   const result: UseFragmentResult<TData> = {
-    data: diff.result,
+    data: diff.result!,
```
Comment (phryneas) [inline src/react/hooks/useFragment.ts:124]: From what I saw in the tests, this will never be `undefined`, but come out as `{}` - I adjusted the types accordingly.
@benjamn, can you please take a look here to make sure I didn't forget about some edge case?

File: src/react/hooks/__tests__/useFragment.test.tsx [hunk 2]
```diff
@@ -349,9 +350,10 @@ describe("useFragment", () => {
         from: { __typename: "Query" },
       });
       expect(complete).toBe(true);
+      assert(!!complete)
```
Comment (phryneas) [inline src/react/hooks/__tests__/useFragment.test.tsx:353]: the `assert` call is here so TypeScript knows that `data` is complete in the JSX.

File: src/react/hooks/useFragment.ts [hunk 3]
```diff
@@ -52,11 +52,17 @@ extends Omit<
...
-export interface UseFragmentResult<TData> {
-  data: TData | undefined;
-  complete: boolean;
-  missing?: MissingTree;
-}
+export type UseFragmentResult<TData> =
+  | {
+      data: TData;
+      complete: true;
+      missing?: MissingTree;
```
Comment (jerelmiller) [inline src/react/hooks/useFragment.ts:59]: I believe `missing` should be `never` here since a `complete` result would imply there are no missing fields.

```suggestion
      missing?: never;
```

File: .changeset/grumpy-tips-know.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
+---
+"@apollo/client": patch
+---
+
+change `UseFragmentResult` to union type
```
Comment (jerelmiller) [inline .changeset/grumpy-tips-know.md:5]: I wonder if we could be a bit more explicit here to better describe the change since this will be on `main` eventually.

```suggestion
More robust types for the `data` property on `UseFragmentResult`. When a partial result is given, the type is now correctly set to `Partial<TData>`.
```

File: src/react/hooks/useFragment.ts [hunk 5]
```diff
@@ -52,11 +52,17 @@ extends Omit<
...
+    }
+  | {
+      data: Partial<TData>;
```
Comment (jerelmiller) [inline src/react/hooks/useFragment.ts:62]: Should we use a `DeepPartial` here since you can give a fragment that could contain nested field selections?

**Review by jerelmiller — COMMENTED** [review]
Had a couple minor suggestions. Thanks for getting this updated!

File: src/react/hooks/useFragment.ts (hunk 5 above)
Comment (phryneas) [reply src/react/hooks/useFragment.ts:62]: Good point. We don't have a `DeepPartial` type at this point yet though - and the one added with #10766 will land in 3.8. We should coordinate around that.

File: src/react/hooks/useFragment.ts (hunk 3 above)
Comment (phryneas) [reply src/react/hooks/useFragment.ts:59]: Good point!

**Comment by phryneas** [comment]
@jerelmiller this now uses `DeepPartial` - once you have given it another glance I'm gonna merge this into the 3.8 branch.

**Review by jerelmiller — APPROVED** [review]
Thanks for updating to `DeepPartial`! These types look great!
